### ISSUE TITLE
Pytorch 2.0: torch.compile() causes tensors to be on a wrong CUDA device when training with DDP

### ISSUE THREAD

**Issue by eldar** [pytorch/pytorch#90723]
### 🐛 Describe the bug

My DDP-based training script fails when using with torch.compile(). The code works fine using either one of the features: that is either DDP without torch.compile() or torch.compile on a single GPU without DDP. I managed to reduce it to a relatively compact reproducible test case:

```python
import os
from tqdm import tqdm
import torch
from torch import nn
import torch.nn.functional as F
import torch.multiprocessing as mp
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP


def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'

    # initialize the process group
    dist.init_process_group("nccl", rank=rank, world_size=world_size)


def cleanup():
    dist.destroy_process_group()


def get_matrix(r):
    if dist.get_rank() == 1:
        print("r.device inside call:", r.device, "rank:", dist.get_rank())

    x, y, z = r.unbind(-1)
    R = torch.stack([-z,  y,
                      z, -x,
                     -y,  x], dim=-1)
    return R.reshape(*x.shape, 3, 2)


class Model(nn.Module):
    def __init__(self):
        super().__init__()

        num_views = 150000
        self.pose_update_r = nn.Parameter(torch.zeros(num_views, 3))
        nn.init.zeros_(self.pose_update_r)

    def forward(self, idx):
        # Triggers error 0
        # r = self.pose_update_r.new_zeros((idx.shape[0], 3))

        # Triggers error 1
        r = F.embedding(idx, self.pose_update_r) # (B, 3)

        # Triggers error 2
        # r = F.embedding(idx, self.pose_update_r, max_norm=0.175) # (B, 3)

        if dist.get_rank() == 1:
            print("r.device before call:", r.device, "rank:", dist.get_rank())
        T = get_matrix(r)
        return T


def train(rank, world_size):
    print(f"Running DDP training on rank {rank}.")
    setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = Model().to(rank)
    model_ddp = DDP(model, device_ids=[rank], find_unused_parameters=True)
    model_ddp = torch.compile(model_ddp)

    for _ in tqdm(range(1000)):
        idx = torch.randint(1000, (8,)).to(rank)
        T = model_ddp(idx)

    cleanup()


if __name__ == "__main__":
    world_size = 2
    mp.spawn(train,
             args=(world_size,),
             nprocs=world_size,
             join=True)
```

These are the print statement outputs I get (full stacktrace is [here](https://gist.github.com/eldar/d515106872f429d17519810284381dd4)):
```
Running DDP training on rank 1.
Running DDP training on rank 0.
r.device before call: cuda:1 rank: 1
r.device inside call: cuda:0 rank: 1
```

Weirdly, the device of tensor `r` in `Model.forward()` inside the rank=1 process changes from `cuda:1` before the function `get_matrix()` is called to `cuda:0` inside `get_matrix()`. `Model.forward` contains three different variants of the code each causing slightly different errors, perhaps they are all related.

### Versions

```
PyTorch version: 2.0.0.dev20221212+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.17

Python version: 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.80.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: Quadro RTX 6000
GPU 1: Quadro RTX 6000
GPU 2: Quadro RTX 6000
GPU 3: Quadro RTX 6000

Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.0rc2
[pip3] torch==2.0.0.dev20221212+cu117
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.15.0.dev20221211+cu117
[conda] numpy                     1.24.0rc2                pypi_0    pypi
[conda] torch                     2.0.0.dev20221212+cu117          pypi_0    pypi
[conda] torchtriton               2.0.0+0d7e753227          pypi_0    pypi
[conda] torchvision               0.15.0.dev20221211+cu117          pypi_0    pypi
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

**Comment by wconstab** [comment]
I have a theory- maybe you can unblock youself and confirm it with a work around.  Can you try setting the default cuda device on process to be the one you'd expect.  Just add this to your init script:`torch.cuda.set_device(rank)`.  (I am assuming here that you intend to use one cuda device and only one cuda device per rank across the whole program).

The theory is that somewhere in dynamo tracing we're not capturing the device faithfully and then we're falling back to the default device at runtime.  We should fix this.

**Comment by eldar** [comment]
Thanks, @wconstab this actually is an obvious thing that I missed - it indeed fixed the issue where the tensor switched the device. However even with this fix I am still getting an error with a second variant of the Model.forward function where `torch.nn.embedding` additionally receives `max_norm` argument:

```
[2022-12-13 21:19:44,254] torch._inductor.graph: [WARNING] Creating implicit fallback for:
  target: aten.embedding_renorm.default                                                                                                                       args[0]: TensorBox(StorageBox(
    Pointwise(
      'cuda',
      torch.float32,
      load(primals_1, i1 + 3 * i0),
      ranges=[150000, 3],                                                                                                                                         origins={clone, primals_1}
    )
  ))
  args[1]: TensorBox(StorageBox(                                                                                                                                InputBuffer(name='primals_2', layout=FixedLayout('cuda', torch.int64, size=[8], stride=[1]))
  ))                                                                                                                                                          args[2]: 0.175
  args[3]: 2.0                                                                                                                                              [2022-12-13 21:19:44,254] torch._inductor.graph: [WARNING] Creating implicit fallback for:
  target: aten.embedding_renorm.default                                                                                                                       args[0]: TensorBox(StorageBox(
    Pointwise(                                                                                                                                                    'cuda',
      torch.float32,                                                                                                                                              load(primals_1, i1 + 3 * i0),
      ranges=[150000, 3],
      origins={clone, primals_1}
    )                                                                                                                                                         ))
  args[1]: TensorBox(StorageBox(                                                                                                                                InputBuffer(name='primals_2', layout=FixedLayout('cuda', torch.int64, size=[8], stride=[1]))
  ))                                                                                                                                                          args[2]: 0.175
  args[3]: 2.0                                                                                                                                              [2022-12-13 21:19:44,260] torch._inductor.ir: [WARNING] Using FallbackKernel: torch.ops.aten.embedding_renorm.default
[2022-12-13 21:19:44,260] torch._inductor.ir: [WARNING] Using FallbackKernel: torch.ops.aten.embedding_renorm.default

  File "/users/eldar/src/hybridrf/train_dynamo_bug.py", line 78, in <module>
    mp.spawn(train,
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 240, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 198, in start_processes
    while not context.join():
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 160, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException:

-- Process 1 terminated with the following error:
Traceback (most recent call last):
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 69, in _wrap
    fn(i, *args)
  File "/users/eldar/src/hybridrf/train_dynamo_bug.py", line 71, in train
    T = model_ddp(idx)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1482, in _call_impl
    return forward_call(*args, **kwargs)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 82, in forward
    return self.dynamo_ctx(self._orig_mod.forward)(*args, **kwargs)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 211, in _fn
    return fn(*args, **kwargs)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1098, in forward
    output = self._run_ddp_forward(*inputs, **kwargs)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1051, in _run_ddp_forward
    return module_to_run(*inputs[0], **kwargs[0])  # type: ignore[index]
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1482, in _call_impl
    return forward_call(*args, **kwargs)
  File "/users/eldar/src/hybridrf/train_dynamo_bug.py", line 42, in forward
    def forward(self, idx):
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 211, in _fn
    return fn(*args, **kwargs)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2343, in forward
    return compiled_fn(full_args)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 887, in g
    return f(*args)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1906, in debug_compiled_function
    return compiled_function(*args)
  File "/users/eldar/apps/anaconda3/envs/ca3d/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1802, in compiled_function
    original_inpt.copy_(updated_inpt)
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Could it be a different issue?

**Comment by eldar** [comment]
Strangely I cannot set `torch._dynamo.config.log_level = logging.DEBUG`

    AttributeError: module 'torch' has no attribute '_dynamo'

I installed Pytorch Nightly via pip.

**Comment by eldar** [comment]
Here is the updated [testcase with the detailed log](https://gist.github.com/eldar/97390726588eecd0cffa615dc194d126).

**Comment by bdhirsh** [comment]
@wconstab actually, @mlazos ran into a similar issue with optimizers. I just wrote up an issue for it: https://github.com/pytorch/pytorch/issues/90844.

Might be worth some discussion first, but I think the fix is pretty simple (I can put up a patch)

**Comment by wconstab** [comment]
> Here is the updated [testcase with the detailed log](https://gist.github.com/eldar/97390726588eecd0cffa615dc194d126).
@eldar thanks for attaching this quickly! Hopefully Brian's fix will help you and we can reinvestigate if not.

**Comment by bdhirsh** [comment]
Here's a patch that you can try @eldar !
https://github.com/pytorch/pytorch/pull/90850

**Comment by eldar** [comment]
Thanks for the quick response! I tried the #90850 and it indeed fixed the remaining bug for me. Now all three variants of the code run. Thank you @bdhirsh @wconstab @nikitaved for the help :) Feel free to close the issue when appropriate.

**Comment by bdhirsh** [comment]
great! Unfortunately, that fix is incomplete, it'll take some more time before we can merge it ([comment](https://github.com/pytorch/pytorch/issues/90844#issuecomment-1353827113))

**Comment by ngimel** [comment]
Closing, the device issue is fixed, if there are remaining issues related to autograd/inplacing, please open a separate issue.
